### PR TITLE
fix: respect arbitrary code execution via $faker runtime expression

### PR DESCRIPTION
**Pull request by DmitryAnansky**
## What/Why/How?

Fixed a remote code execution vulnerability where a crafted $faker expression in
an Arazzo description could execute arbitrary JavaScript during respect runs.
$faker expressions are now resolved against the known faker object with literal-only
argument parsing, blocking prototype-chain escapes.

Thanks to Hamza Haroon (GitHub: @thegr1ffyn) for the report.

## Testing

Internally => https://github.com/Redocly/redocly/pull/23998

## Check yourself

- [x] This PR follows the [contributing guide](https://github.com/Redocly/redocly-cli/blob/main/CONTRIBUTING.md#pull-request-guidelines) 
- [x] All new/updated code is covered by tests
- [ ] Core code changed? - Tested with other Redocly products (internal contributions only)
- [ ] New package installed? - Tested in different environments (browser/node)
- [x] Documentation update has been considered 

## Security

- [x] The security impact of the change has been considered 
- [x] Code follows company security practices and guidelines

---

> [!NOTE]
> **High Risk**
> This closes a remote code execution path in Respect’s expression resolver; the change is security-critical but narrows behavior to safe faker access rather than broadening attack surface.
> 
> **Overview**
> **Security fix:** `$faker` runtime expressions in Arazzo descriptions no longer run through dynamic `new Function` evaluation, which allowed arbitrary JavaScript execution (e.g. via `constructor` / bracket-notation escapes).
> 
> **`getFakeData` now** walks only **own properties** on the real faker instance: pointers are split with `splitFakerPointer` (dots inside calls/strings preserved), each segment is resolved with `resolveFakerSegment`, and names like `constructor`, `__proto__`, and `prototype` are blocked. Function arguments are parsed by a **literal-only** `FakerArgumentParser` (numbers, strings, simple option objects)—not executed as JS.
> 
> Tests add exploit regression cases and cover argument parsing edge cases (floats, negatives, malformed input). A patch changeset documents the fix for `@redocly/respect-core`.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 81a2e988111e863af4b542dd08f11a519bf6e5ad. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: packages/respect-core/src/modules/__tests__/config-parser/get-value-from-context.test.ts [hunk 1]
```diff
@@ -223,6 +223,125 @@ describe('getValueFromContext', () => {
...
+    expect(received).toEqual([{ min: 0.5, max: 1.5 }]);
+  });
+
+  it('parses arrays containing dotted strings', () => {
+    expect(call("$faker.helpers.arrayElement(['example.com', 'test.org'])")).toEqual('ok');
+    expect(received).toEqual([['example.com', 'test.org']]);
+  });
+
+  it('parses nested objects', () => {
+    expect(call('$faker.custom.echo({ a: { b: 1 }, c: [2, 3] })')).toEqual('ok');
```
Comment (tatomyr) [inline packages/respect-core/src/modules/__tests__/config-parser/get-value-from-context.test.ts]: What is custom.echo? Do we even need it? Wouldn't it be easier to simply allow known used words? In the end of the day, we're aiming at simple primitives like strings/emails/numbers etc..

Comment (DmitryAnansky) [reply packages/respect-core/src/modules/__tests__/config-parser/get-value-from-context.test.ts]: Its just a dummy function we defined in test mock so we could only focus on returned parameters  result.
```
    const echo = (...args: unknown[]) => {
      received.push(...args);
      return 'ok';
    };
```
We can also replace it with real $faker funcitons.

Comment (DmitryAnansky) [reply packages/respect-core/src/modules/__tests__/config-parser/get-value-from-context.test.ts]: updated, please take a look.

File: .changeset/slimy-beers-fry.md [hunk 2]
```diff
@@ -0,0 +1,7 @@
+---
+'@redocly/respect-core': patch
+---
+
+Fixed a remote code execution vulnerability where a crafted $faker expression in
+an Arazzo description could execute arbitrary JavaScript during respect runs.
+Thanks to Hamza Haroon (GitHub: @thegr1ffyn) for the report.
```
Comment (JLekawa) [inline .changeset/slimy-beers-fry.md]:
```suggestion
Fixed a remote code execution vulnerability where a crafted `$faker` expression in an Arazzo description could execute arbitrary JavaScript during Redocly Respect runs.
Reported by Hamza Haroon (GitHub: @thegr1ffyn).
```

**Comment by thegr1ffyn** [comment]
Do let me know if you need to test the updated version against the PoC.

File: packages/respect-core/src/modules/__tests__/config-parser/get-value-from-context.test.ts [hunk 3]
```diff
@@ -223,6 +223,98 @@ describe('getValueFromContext', () => {
...
+describe('getFakeData argument parsing', () => {
+  let ctx: TestContext;
+
+  beforeEach(() => {
+    ctx = {
+      $faker: createFaker(),
+    } as unknown as TestContext;
+  });
+
+  const call = (value: string) => getValueFromContext({ value, ctx, logger });
```
Comment (tatomyr) [inline packages/respect-core/src/modules/__tests__/config-parser/get-value-from-context.test.ts]: Please do not use wrappers.

File: packages/respect-core/src/modules/__tests__/config-parser/get-value-from-context.test.ts [hunk 4]
```diff
@@ -223,6 +223,98 @@ describe('getValueFromContext', () => {
...
+      'example.org'
+    );
+  });
+
+  it('tolerates trailing commas', () => {
+    expect(call('$faker.number.integer({ min: 5, max: 5, })')).toEqual(5);
+  });
+
+  it('strips prototype-polluting keys from parsed argument objects', () => {
+    expect(call('$faker.number.integer({ __proto__: { polluted: 1 }, min: 5, max: 5 })')).toEqual(
```
Comment (tatomyr) [inline packages/respect-core/src/modules/__tests__/config-parser/get-value-from-context.test.ts]: Please make sure your tests are failing without the fix.

**Comment by DmitryAnansky** [comment]
> Do let me know if you need to test the updated version against the PoC.

That would be great. You can use the snapshot version:
@redocly/cli@0.0.0-snapshot.1781521186

**Comment by thegr1ffyn** [comment]
> > Do let me know if you need to test the updated version against the PoC.
> 
> That would be great. You can use the snapshot version: @redocly/cli@0.0.0-snapshot.1781521186

Ill get back to you after the test on @redocly/cli@0.0.0-snapshot.1781521186 in a while

File: packages/respect-core/src/modules/__tests__/config-parser/get-value-from-context.test.ts (hunk 4 above)
Comment (DmitryAnansky) [reply packages/respect-core/src/modules/__tests__/config-parser/get-value-from-context.test.ts]: All 3 failing on main branch with Pwned or Func context:
```
     × should not execute arbitrary code via the $faker constructor escape 3ms
     × should not execute arbitrary code via bracket-notation $faker payloads 0ms
     × should not resolve prototype-chain properties via $faker 0ms
```

**Comment by thegr1ffyn** [comment]
The RCE is fixed in this snapshot (@redocly/cli@0.0.0-snapshot.1781521186) and the fix is sound and well-tested. 
@DmitryAnansky

**Review by JLekawa — APPROVED** [review]

**Comment by thegr1ffyn** [comment]
Just a kind request that when you merge and the latest version is out, please create a GitHub Advisory for this and request for CVE assignment. @JLekawa @DmitryAnansky

File: packages/respect-core/src/modules/__tests__/config-parser/get-value-from-context.test.ts [hunk 5]
```diff
@@ -223,6 +223,109 @@ describe('getValueFromContext', () => {
...
+  });
+});
+
+describe('getFakeData argument parsing', () => {
+  let ctx: TestContext;
+
+  beforeEach(() => {
+    ctx = {
+      $faker: createFaker(),
+    } as unknown as TestContext;
```
Comment (tatomyr) [inline packages/respect-core/src/modules/__tests__/config-parser/get-value-from-context.test.ts]: This doublecasting doesn't make sense.
```suggestion
    } as any;
```

**Review by tatomyr — APPROVED** [review]

**Comment by DmitryAnansky** [comment]
@thegr1ffyn 
We will provide an update here once a fix has been released.
Thank you for reporting the issue.

**Comment by thegr1ffyn** [comment]
Thanks, Ill be waiting.
Meanwhile, I reported some more vulnerabilities in the email as well. Feel free to check them in your time and ping me, Ill be here for any technical support needed.
@DmitryAnansky

**Comment by DmitryAnansky** [comment]
> Thanks, Ill be waiting. Meanwhile, I reported some more vulnerabilities in the email as well. Feel free to check them in your time and ping me, Ill be here for any technical support needed. @DmitryAnansky

@thegr1ffyn 
Fix is available in `v.2.33.0`.
We will create a CVE and provide an update in this PR once it has been published.
Thank you.

**Comment by thegr1ffyn** [comment]
Hello @tatomyr @JLekawa @DmitryAnansky 
Since the vulnerability is patched and updated version is released, can you please publish GitHub advisory with the details I mentioned in the email and request for CVE?

**Comment by DmitryAnansky** [comment]
> Hello @tatomyr @JLekawa @DmitryAnansky Since the vulnerability is patched and updated version is released, can you please publish GitHub advisory with the details I mentioned in the email and request for CVE?

It takes some time to complete due to our internal processes, but we will notify you as soon as it is ready. Thank you for your patience.

**Comment by thegr1ffyn** [comment]
Hello team @JLekawa @Daryna-del @DmitryAnansky @tatomyr 

Its been a while since I reported this vulnerability. I kindly request to release Github Advisory for this alongwith Requesting a CVE on the advisory once published.

**Comment by tatomyr** [comment]
@thegr1ffyn we've already requested the CVEs. We'll notify you once GitHub provides them, and the Advisories are published.

**Comment by tatomyr** [comment]
@thegr1ffyn published the advisories: https://github.com/Redocly/redocly-cli/security/advisories/GHSA-xw2f-5386-m542, https://github.com/Redocly/redocly-cli/security/advisories/GHSA-657c-g7qc-r9j2. Thanks for your report.
